### PR TITLE
regex querying to LLM interaction

### PR DESCRIPTION
**Pull request by itsmeashutosh43**
I have added a regex_conditioner module which conditions the next word based on regex pattern. Here are the highlights of the implementation

- Output tokens are generated one at a time instead of in a batch
- For tokens that does not match the regex pattern, we set its logit bias to negative infinity
- I have added some tests at tests/test_conditional_generation.py. Due to computational constraints I am not able to test it with large models and bloom is the largest model that I could test. It should be noted that bloom is popularly used for completion therefore my test cases are for completion tasks.
- Most parts of my implementation is copied / modified / inspired from [here](https://github.com/r2d4/rellm)

### CODE REVIEW

**Comment by mmirman** [comment]
What exactly was copied/what is the difference from reLLM?  If its similar enough can we include it with a wrapper over a subtree?  Is rellm on pypy yet?

**Comment by itsmeashutosh43** [comment]
Frankly, in terms of functionality, there is no difference. I have added some async functions and restructured the flow to make it more efficient and amenable to this repository. Initially, I had thought of an ad-hoc solution with a wrapper, but that would only work for regex conditioning. Instead, I wanted to make it more open to handlebars and other ways through which we can condition the LLM.

Please let me know if you have more output templates in your roadmap. However, since there's no functional difference and if you feel a wrapper-based solution for regex will be more straightforward, I don't have much to do this weekend. I can give it a try."

**Comment by mmirman** [comment]
Making it a wrapper until there is a difference in functionionality means we can utilize rellm's security fixes and updates significantly more easily.  Also less trouble legally.

**Comment by mmirman** [comment]
But yeah, since the correct solution here is just a wrapper let's broaden the scope of this ticket to some context sensitive grammar.  Up for debate which grammar makes sense.  Maybe even regular expressions with variable references (although its nonstandard regex syntax, I'd guess its easy to understand)?

**Comment by itsmeashutosh43** [comment]
Hmm, regular expressions with variable references sounds interesting. Can you give some examples?

**Comment by mmirman** [comment]
https://www.rexegg.com/regex-capture.html

**Comment by mmirman** [comment]
rellm might already handle this if they are just using an existing regex implemention though :-)

**Review by cartazio — COMMENTED** [review]
i think that this implementation wont work on responses that are a prefix of a valid accepting regex match, which is a failure in specification on our part

File: pyvenv.cfg
Comment (cartazio) [inline pyvenv.cfg:1]: this shouldnt be in your commits

File: settings.default.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Default settings for your application
 BIG_MODEL = "chat_gpt"
 PORT = 3002
-SMALL_MODEL = "pythia"
+SMALL_MODEL = "opt"
```
Comment (cartazio) [inline settings.default.toml:4]: why this change?

File: src/llm_vm/client.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
     default_big_model= conf.settings.big_model
 
 if conf.settings.small_model is not  None:
-    default_small_model= "pythia"
+    default_small_model= "opt"
```
Comment (cartazio) [inline src/llm_vm/client.py:16]: why this change?

File: src/llm_vm/client.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def __init__(self, big_model = default_big_model, small_model =default_small_mod
 
         #     return self.teacher.generate(prompt, max_len,**kwargs)
 
-        def CALL_SMALL(prompt, max_len=256, **kwargs):
+        def CALL_SMALL(prompt, max_len=40, **kwargs):
```
Comment (cartazio) [inline src/llm_vm/client.py:64]: why this change?

File: src/llm_vm/conditioner/regex_conditioner.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+class LogitsMask(LogitsProcessor):
+    """
+    LogitsMask is a LogitsProcessor that masks logits for tokens that are 
+    not in the allowed token ids set.
+    """
+    def __init__(self, allowed_token_ids):
+        self.allowed_token_ids = set(allowed_token_ids)
+
+    def __call__(self, input_ids, scores):
+        mask = np.ones_like(scores) * -1e10
```
[the remark this answers is not in the document]
Comment (itsmeashutosh43) [reply src/llm_vm/conditioner/regex_conditioner.py:65]: Log likelihood of negative infinity(likelihood 0) if we don't want tokens that don't match the pattern.

File: settings.default.toml (hunk 1 above)
Comment (itsmeashutosh43) [reply settings.default.toml:4]: sorry my bad

File: src/llm_vm/client.py (hunk 2 above)
Comment (itsmeashutosh43) [reply src/llm_vm/client.py:16]: sorry my bad

File: src/llm_vm/client.py (hunk 3 above)
Comment (itsmeashutosh43) [reply src/llm_vm/client.py:64]: sorry my bad

Comment (mmirman) [reply src/llm_vm/client.py:64]: Don't sweat it!  It's just an annoying necessity to keep pull requests minimal and local for bookkeeping 😃

**Comment by mmirman** [comment]
Remember to join in on https://discord.gg/DP2B2S2A7m?event=1138173908865003560

**Comment by cartazio** [comment]
@abhigya-sodani  is doing some work adapting a new algorithmic state of the art for this sortah inference into the code base, coordinate with him how to help out
